### PR TITLE
Issue: 260 - Paramaterize inspection links

### PR DESCRIPTION
**Pull request by ghost**
**Fixes issue #260**:

**Description of the changes being introduced by the pull request**:
Added an optional boolean parameter to run_all_inspections in verifylib.py which determines whether or not a metadata link file is written. When true, the file is written to the client's cwd. When false, no file is written. Set to true by default, since that was the behavior before this change.

**Please verify and check that the pull request fulfills the following
requirements**:

- [X] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [X] Tests have been added for the bug fix or new feature
- [X] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by adityasaky** [comment]
Thanks for taking this issue up, @alexsmith15! One of us will be along shortly to take a look.

File: tests/test_verifylib.py [hunk 1]
```diff
@@ -146,6 +146,38 @@ def test_inspection_fail_with_non_zero_retval(self):
     with self.assertRaises(BadReturnValueError):
       run_all_inspections(layout)
 
+  def test_inspection_persistence_false(self):
+    """Test metadata link file persistence"""
+    scripts_directory = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "scripts")
+
+    layout = Layout.read({
```
Comment (lukpueh) [inline tests/test_verifylib.py]: You could just use `self.layout` instead of redefining one here.

File: tests/test_verifylib.py [hunk 2]
```diff
@@ -146,6 +146,38 @@ def test_inspection_fail_with_non_zero_retval(self):
...
+    })
+    run_all_inspections(layout)
+    self.assertTrue(os.path.exists("touch-foobar.link"))
+
+  def test_inspection_persistence_true(self):
+    """Test metadata link file non-persistence"""
+    scripts_directory = os.path.join(
+        os.path.dirname(os.path.realpath(__file__)), "scripts")
+
+    layout = Layout.read({
```
Comment (lukpueh) [inline tests/test_verifylib.py]: See comment above about `self.layout`

File: in_toto/verifylib.py [hunk 3]
```diff
@@ -1423,6 +1432,12 @@ def in_toto_verify(layout, layout_key_dict, link_dir_path=".",
 
     step_name (optional): A name assigned to the returned link. This is mostly
         useful during recursive sublayout verification.
+    
+    persist_inspection_links: (optional)
+            A boolean that determines whether or not link metadata files for
+            inspection are written to cwd.
+            When true, the link inspection file is written to cwd.
+            When false, the file is not written.
```
Comment (lukpueh) [inline in_toto/verifylib.py]: Minor nit: Can you un-indent one level to be consistent with the other arguments? Also, I think the the first two lines of the description should be enough to understand how to use the argument.

File: in_toto/verifylib.py [hunk 4]
```diff
@@ -167,7 +167,7 @@ def load_links_for_layout(layout, link_dir_path):
   return steps_metadata
 
 
-def run_all_inspections(layout):
+def run_all_inspections(layout, persist_inspection_links=True):
```
Comment (lukpueh) [inline in_toto/verifylib.py]: Minor nit: I think it's okay to make this a non-optional argument here. 

Although not identified as such by leading underscore(s), `run_all_inspections`  is really more of an internal function that's only called by `in_toto_verify`, which passes `persist_inspection_links` no matter what.

**Review by lukpueh — COMMENTED** [review]
Great work, @alexsmith15, and thanks for taking a stab at #260! 

@trishankatdatadog, @radu-matei, since you signaled interest in the original issue, would you mind giving feedback whether an on/off solution fits our needs? Or if we also/instead want to allow a custom path for inspection links?

File: tests/test_verifylib.py [hunk 5]
```diff
@@ -146,6 +146,38 @@ def test_inspection_fail_with_non_zero_retval(self):
     with self.assertRaises(BadReturnValueError):
       run_all_inspections(layout)
 
+  def test_inspection_persistence_false(self):
```
Comment (trishankatdatadog) [inline tests/test_verifylib.py:149]:
```suggestion
  def test_inspection_persistence_true(self):
```

File: tests/test_verifylib.py [hunk 6]
```diff
@@ -146,6 +146,38 @@ def test_inspection_fail_with_non_zero_retval(self):
...
+        "steps": [],
+        "inspect": [{
+          "name": "touch-foobar",
+          "run": ["python", os.path.join(scripts_directory, "touch"), "bar"],
+        }]
+    })
+    run_all_inspections(layout)
+    self.assertTrue(os.path.exists("touch-foobar.link"))
+
+  def test_inspection_persistence_true(self):
```
Comment (trishankatdatadog) [inline tests/test_verifylib.py]:
```suggestion
  def test_inspection_persistence_false(self):
```

**Review by trishankatdatadog — CHANGES_REQUESTED** [review]
> @trishankatdatadog, @radu-matei, since you signaled interest in the original issue, would you mind giving feedback whether an on/off solution fits our needs? Or if we also/instead want to allow a custom path for inspection links?

Thanks for your work, @alexsmith15, and for pinging us, @lukpueh! I think this is good enough for now, aside for some minor changes. What do you think, @radumatei?

**Comment by trishankatdatadog** [comment]
> Thanks for your work, @alexsmith15, and for pinging us, @lukpueh! I think this is good enough for now, aside for some minor changes. What do you think, @RaduMatei?

Especially since the Python reference implementation still implicitly assumes storing the layout, links, and pubkeys all in the same directory anyway, right? (Which is why we need to `ALLOW` for these files as materials to the inspection.)

**Comment by lukpueh** [comment]
> Especially since the Python reference implementation still implicitly assumes storing the layout, links, and pubkeys all in the same directory anyway, right? (Which is why we need to ALLOW for these files as materials to the inspection.)

in-toto-run/verify cli/api allows or requires you to pass paths for all these.

**Comment by trishankatdatadog** [comment]
> in-toto-run/verify cli/api allows or requires you to pass paths for all these.

Then we're set, I think.

**Comment by SantiagoTorres** [comment]
Awesome stuff @alexsmith15 ! could you please make sure to sign-off your commits and re-push? :)

**Comment by ghost** [comment]
Hey thanks for the feedback and sorry for the delay! Yes I will get that fixed by the end of the day today!

**Comment by ghost** [comment]
Re-did commits with sign-off and addressed the changes that were suggested. Let me know if I need to make other changes or if I missed someting!

File: tests/test_verifylib.py [hunk 7]
```diff
@@ -144,7 +144,18 @@ def test_inspection_fail_with_non_zero_retval(self):
         }]
     })
     with self.assertRaises(BadReturnValueError):
-      run_all_inspections(layout)
+      run_all_inspections(layout, True)
+
+  def test_inspection_persistence_false(self):
+    """Test metadata link file non-persistence"""
+    os.remove("touch-bar.link")
```
Comment (lukpueh) [inline tests/test_verifylib.py:151]: It's a good idea to make sure that there is no `touch-bar.link` before running the inspections, but blindly removing it if it indeed does not exist fails the tests. I suggest to just assert that it does not exist.

File: in_toto/verifylib.py [hunk 8]
```diff
@@ -219,11 +223,12 @@ def run_all_inspections(layout):
...
     inspection_links_dict[inspection.name] = link
 
+    # If the client requests persistent inspection links,
     # Dump the inspection link file for auditing
     # Keep in mind that this pollutes the verifier's (client's) filesystem.
-    filename = in_toto.models.link.FILENAME_FORMAT_SHORT.format(
-        step_name=inspection.name)
-    link.dump(filename)
+    if persist_inspection_links:
+      filename = FILENAME_FORMAT_SHORT.format(step_name=inspection.name)
```
Comment (lukpueh) [inline in_toto/verifylib.py:230]: The direct import of `FILENAME_FORMAT_SHORT` was removed in a recent PR (#379) please use the full path.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Many thanks for resolving the conflicts and addressing the review comments! 🎉 

Please always make sure that the test pass before you request a review  (see [review guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/master/code-review.md)) and please also don't include merge commits in your PRs (see [git history guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/master/git-history.md)). :)

**Comment by adityasaky** [comment]
@kristelfung's takeover PR #392 has now been merged.

**Pull request closed without merging**